### PR TITLE
feat: add preliminary map tile creation

### PR DESCRIPTION
**Pull request by edparris**
This PR gives the tile server the ability to create orthorectified tiles from imagery. New classes have been added to define the common WebMercatorQuad tile set and a new function on the GDALTileFactory can be used to create those tiles directly from an image. This processing uses the sensor model and NumPy to calculate the relationships between the original image and the desired tile then the OpenCV remap function does the heavy lifting of manipulating the pixels. Preliminary performance tests were run by integrating this function into the osml-tile-server and running a load test to exercise the map API. The response time was similar to the the regular tile cutting routines so no unusual performance impacts are expected. 

### Notes
- This interface has been flagged as experimental and we may choose to update the public API before the next release.
- This code adds a new dependency on [OpenCV](https://opencv.org/).

### Checklist

Before you submit a pull request, please make sure you have the following:
- [x] Code changes are compact and well-structured to facilitate easy review
- [x] Changes are documented in the README.md and other relevant documentation pages
- [x] PR title and description accurately reflect the changes and are detailed enough for historical tracking
- [x] PR contains tests that cover all new code and the code has been manual tested
- [x] All new dependencies are declared (if any), and no unnecessary libraries are added
- [x] Performance impacts (if any) of the changes are evaluated and documented
- [x] Security implications of the changes (if any) are reviewed and addressed
- [x] I have read the [Contributing Guidelines](../CONTRIBUTING.md) and agree to follow the [Code of Conduct](../CODE_OF_CONDUCT.md)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/aws/osml/image_processing/__init__.py [hunk 1]
```diff
@@ -93,6 +122,17 @@
 """
 
 from .gdal_tile_factory import GDALTileFactory
+from .map_tileset import MapTile, MapTileId, MapTileSet
+from .map_tileset_factory import MapTileSetFactory, WellKnownMapTileSet
 from .sar_complex_imageop import histogram_stretch, quarter_power_image
 
-__all__ = ["GDALTileFactory", "histogram_stretch", "quarter_power_image"]
+__all__ = [
```
Comment (drduhe) [inline src/aws/osml/image_processing/__init__.py:129]: Should we go back and standardize all our inits to use this __all__ pattern?

File: src/aws/osml/image_processing/gdal_tile_factory.py [hunk 2]
```diff
@@ -135,7 +139,199 @@ def create_encoded_tile(
...
+
+        # TODO: Formats other than PNG?
+        is_success, image_bytes = cv2.imencode(".png", output_tile_pixels)
+        if is_success:
+            return image_bytes
+        else:
+            return None
+
+    def _read_from_rlevel_as_array(
+        self, scaled_bbox: tuple[int, int, int, int], r_level: int, band_numbers: List[int] = None
```
Comment (drduhe) [inline src/aws/osml/image_processing/gdal_tile_factory.py:264]: You can make this List[int] = None Optional instead to match convention.

File: src/aws/osml/image_processing/map_tileset_factory.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    @staticmethod
+    def get_for_id(tile_matrix_set_id: str) -> Optional[MapTileSet]:
+        """
+        Constructs a tile set matching the requested id.
+
+        :param tile_matrix_set_id: the tile set id
+        :return: the TileSet or None if not available
+        """
+        if tile_matrix_set_id == WellKnownMapTileSet.WEB_MERCATOR_QUAD.value:
```
Comment (drduhe) [inline src/aws/osml/image_processing/map_tileset_factory.py:34]: Should the tile size me an initialization argument perhaps? Might that change?

**Review by drduhe — APPROVED** [review]

File: src/aws/osml/image_processing/map_tileset_factory.py (hunk 3 above)
Comment (edparris) [reply src/aws/osml/image_processing/map_tileset_factory.py:34]: No; the tile size is specified by the tile set. In this specific instance the "WebMercatorQuad" tile set specifies 256x256 tiles and "WebMercatorQuadx2" is almost identical except each tile is 512x512. In the WebMercatorQuadMapTileSet implementation I made tile size an argument so the same code can work for either but here at the factory when an end user is asking for a specific tile set by ID then the tile size is implied.

File: src/aws/osml/image_processing/gdal_tile_factory.py (hunk 2 above)
Comment (edparris) [reply src/aws/osml/image_processing/gdal_tile_factory.py:264]: Yep. Will fix.

File: src/aws/osml/image_processing/__init__.py (hunk 1 above)
Comment (edparris) [reply src/aws/osml/image_processing/__init__.py:129]: The `__all__` list in an init.py defines the set of module names that are imported when someone imports *. The imagery toolkit modules are already making use of this as a way to explicitly define what is part of the public library API and what is not. 

In this PR if a consumer writes `from aws.osml.image_processing import *` then they will get the names listed here which deliberately excludes the WebMercatorQuadTileSet implementation of MapTileSet. The idea is that users should create these well known tile sets using the factory / enumerated IDs in WellKnownMapTileSet and never need to import or instantiate the concrete implementations directly.

**Review by RanbirAulakh — APPROVED** [review]

**Pull request merged**